In [19]:
import re

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 1: Trebuchet?! ---</h2><p>Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.</p>
<p>You've been doing this long enough to know that to restore snow operations, you need to check all <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by solving puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
<p>You try to ask why they can't just use a <a href="/2015/day/1">weather machine</a> ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") <span title="My hope is that this abomination of a run-on sentence somehow conveys the chaos of being hastily loaded into a trebuchet.">and</span> hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a <a href="https://en.wikipedia.org/wiki/Trebuchet" target="_blank">trebuchet</a> ("please hold still, we need to strap you in").</p>
<p>As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been <em>amended</em> by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.</p>
<p>The newly-improved calibration document consists of lines of text; each line originally contained a specific <em>calibration value</em> that the Elves now need to recover. On each line, the calibration value can be found by combining the <em>first digit</em> and the <em>last digit</em> (in that order) to form a single <em>two-digit number</em>.</p>
<p>For example:</p>
<pre><code>1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
</code></pre>
<p>In this example, the calibration values of these four lines are <code>12</code>, <code>38</code>, <code>15</code>, and <code>77</code>. Adding these together produces <code><em>142</em></code>.</p>
<p>Consider your entire calibration document. <em>What is the sum of all of the calibration values?</em></p>
</article>


In [20]:
test = [("1abc2", 12), ("pqr3stu8vwx", 38), ("a1b2c3d4e5f", 15), ("treb7uchet", 77)]

document = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"""

total = 142

In [21]:
def sumAllCaqlibrated(document: str) -> int:
    som = 0
    for line in document.splitlines():
        first, last = None, None
        i, n = 0, len(line)
        while i < n and (first is None or last is None):
            if first is None and line[i].isdigit():
                first = line[i]
            if last is None and line[n - i - 1].isdigit():
                last = line[n - i - 1]
            i += 1
        som += int(first + last)
    return som


assert sumAllCaqlibrated(document) == total

In [22]:
with open("../input/day1.txt") as f:
    print(sumAllCaqlibrated(f.read()))

54644


<link href="style.css" rel="stylesheet"></link>
<p>Your puzzle answer was <code>54644</code>.</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Your calculation isn't quite right. It looks like some of the digits are actually <em>spelled out with letters</em>: <code>one</code>, <code>two</code>, <code>three</code>, <code>four</code>, <code>five</code>, <code>six</code>, <code>seven</code>, <code>eight</code>, and <code>nine</code> <em>also</em> count as valid "digits".</p>
<p>Equipped with this new information, you now need to find the real first and last digit on each line. For example:</p>
<pre><code>two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
</code></pre>
<p>In this example, the calibration values are <code>29</code>, <code>83</code>, <code>13</code>, <code>24</code>, <code>42</code>, <code>14</code>, and <code>76</code>. Adding these together produces <code><em>281</em></code>.</p>
<p><em>What is the sum of all of the calibration values?</em></p>
</article>


In [23]:
document = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"""


document2 = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"""

In [24]:
DIGITS = re.compile(
    r"(?=(one|two|three|four|five|six|seven|eight|nine|[0-9]))"  # for overlapping
)

assert DIGITS.findall("sevenineeighthree") == [
    "seven",
    "nine",
    "eight",
    "three",
]  # test for overlapping


def convert(s: str) -> int:
    match s:
        case "one" | "1":
            return "1"
        case "two" | "2":
            return "2"
        case "three" | "3":
            return "3"
        case "four" | "4":
            return "4"
        case "five" | "5":
            return "5"
        case "six" | "6":
            return "6"
        case "seven" | "7":
            return "7"
        case "eight" | "8":
            return "8"
        case "nine" | "9":
            return "9"
        case _:
            raise ValueError(f"{s=}")


def sumAllCaqlibrated_II(document: str) -> int:
    som = 0
    for line in document.splitlines():
        matches = DIGITS.findall(line)
        first, last = convert(matches[0]), convert(matches[-1])
        som += int(first + last)
    return som


assert sumAllCaqlibrated_II(document) == 281
assert sumAllCaqlibrated_II(document2) == total

In [25]:
with open("../input/day1.txt") as f:
    assert sumAllCaqlibrated_II(f.read()) == 53348

<main>

<p>Your puzzle answer was <code>53348</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2023">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="1/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Trebuchet%3F%21%22+%2D+Day+1+%2D+Advent+of+Code+2023&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F1&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Trebuchet%3F%21%22+%2D+Day+1+%2D+Advent+of+Code+2023+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F1';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
